sklearn/compose/_column_transformer.py

```diff
@@ -453,6 +459,10 @@ def _get_feature_name_out_for_transformer(
                 f"Transformer {name} (type {type(trans).__name__}) does "
                 "not provide get_feature_names_out."
             )
+        if isinstance(column, slice) and _determine_key_type(column) == "str":
```

Comment:
Same here:

```python
column_indices = self._transformer_to_input_indices[name]
names = feature_names_in[column_indices]
return trans.get_feature_names_out(names)
```

and remove the if statements.

Reply:
@thomasjpfan When I was writing this, I was not sure if keeping `column` severed another purpose. Completely depending only on `_transformer_to_input_indices` will simplify the logic even more as in the latest PR.

Reply:
Since `_get_feature_name_out_for_transformer` is private, we can change the API if we need to. We need to make sure the test pass for the public `get_feature_names_out` API.

Although in this case, the `column` is still used in:

```python
if trans == "drop" or _is_empty_column_selection(column):
``` 

I think we can keep this line here for now, so this PR is easier to review. In the future, we can open an follow up PR to refactor `_get_feature_name_out_for_transformer`.